### ISSUE TITLE
[Perf] Regressions in math based array benchmarks

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#52316]
### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [ac82799250fe42c8ba2941aef487aca94ecf04cc](https://github.com/dotnet/runtime/commit/ac82799250fe42c8ba2941aef487aca94ecf04cc)
Compare | [1c9e2009d9942a6df31daad8c912c68446856fa2](https://github.com/dotnet/runtime/commit/1c9e2009d9942a6df31daad8c912c68446856fa2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/ac82799250fe42c8ba2941aef487aca94ecf04cc...1c9e2009d9942a6df31daad8c912c68446856fa2)

### Regressions in ByteMark

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[BenchLUDecomp - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04/ByteMark.BenchLUDecomp.html>) | 967.28 ms | 1.34 secs | 1.38 | 0.00 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/ByteMark_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/ByteMark.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'ByteMark*'
```
<details>

### Payloads

[Baseline](<>)
[Compare](<>)

#### ByteMark.BenchLUDecomp

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [ac82799250fe42c8ba2941aef487aca94ecf04cc](https://github.com/dotnet/runtime/commit/ac82799250fe42c8ba2941aef487aca94ecf04cc)
Compare | [1c9e2009d9942a6df31daad8c912c68446856fa2](https://github.com/dotnet/runtime/commit/1c9e2009d9942a6df31daad8c912c68446856fa2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/ac82799250fe42c8ba2941aef487aca94ecf04cc...1c9e2009d9942a6df31daad8c912c68446856fa2)

### Regressions in Benchstone.BenchI.Array2

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[Test - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04/Benchstone.BenchI.Array2.Test.html>) | 357.67 ms | 566.68 ms | 1.58 | 0.01 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/Benchstone.BenchI.Array2_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/Benchstone.BenchI.Array2.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Benchstone.BenchI.Array2*'
```
<details>

### Payloads

[Baseline](<>)
[Compare](<>)

#### Benchstone.BenchI.Array2.Test

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [ac82799250fe42c8ba2941aef487aca94ecf04cc](https://github.com/dotnet/runtime/commit/ac82799250fe42c8ba2941aef487aca94ecf04cc)
Compare | [1c9e2009d9942a6df31daad8c912c68446856fa2](https://github.com/dotnet/runtime/commit/1c9e2009d9942a6df31daad8c912c68446856fa2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/ac82799250fe42c8ba2941aef487aca94ecf04cc...1c9e2009d9942a6df31daad8c912c68446856fa2)

### Regressions in BenchmarksGame.FannkuchRedux_2

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[RunBench - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04/BenchmarksGame.FannkuchRedux_2.RunBench(n%3a%2010%2c%20expectedSum%3a%2073196).html>) | 126.48 ms | 143.40 ms | 1.13 | 0.00 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/BenchmarksGame.FannkuchRedux_2_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/BenchmarksGame.FannkuchRedux_2.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'BenchmarksGame.FannkuchRedux_2*'
```
<details>

### Payloads

[Baseline](<>)
[Compare](<>)

#### BenchmarksGame.FannkuchRedux_2.RunBench(n: 10, expectedSum: 73196)

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [ac82799250fe42c8ba2941aef487aca94ecf04cc](https://github.com/dotnet/runtime/commit/ac82799250fe42c8ba2941aef487aca94ecf04cc)
Compare | [1c9e2009d9942a6df31daad8c912c68446856fa2](https://github.com/dotnet/runtime/commit/1c9e2009d9942a6df31daad8c912c68446856fa2)
Diff | [Diff](https://github.com/dotnet/runtime/compare/ac82799250fe42c8ba2941aef487aca94ecf04cc...1c9e2009d9942a6df31daad8c912c68446856fa2)

### Regressions in SciMark2.kernel

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[benchSparseMult - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04/SciMark2.kernel.benchSparseMult.html>) | 512.50 ms | 610.62 ms | 1.19 | 0.00 | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/SciMark2.kernel_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_03_2021/refs/heads/main_x64_ubuntu%2018.04_Regression/SciMark2.kernel.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'SciMark2.kernel*'
```
<details>

### Payloads

[Baseline](<>)
[Compare](<>)

#### SciMark2.kernel.benchSparseMult

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by DrewScoggins** [comment]
This change seems to be the cause, https://github.com/dotnet/runtime/pull/51901

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib

**Comment by BruceForstall** [comment]
The ludcmp issue is an LSRA spill cost issue, detailed in #53703.

The Array2 regression is due to unfortunate loop alignment of the innermost nested loop (of a 4 deep loop nest): in the new, slow case, this innermost loop spans 3 32-byte chunks, whereas before, it only spanned 2. The JIT aligns the loop up to 16 bytes, but that's not enough here. Oddly, I can easily repro this regression on my Linux x64 box, but not my Windows x64 box (where the test didn't show any regression in the perf lab, and in fact has been incredibly consistent, compared to the Linux one, which has moved around a bit). If I set `COMPlus_JitAlignLoopAdaptive=0`, the performance swaps: the previously "slow" version is across 2 chunks and the previously fast version is across 3 chunks... and, their times swap.

**Comment by kunalspathak** [comment]
So, I just generated disassembly of before (your loop weight change) vs. after, and I see things other way round.

In below screenshot, left is before and right is after. Earier, we were aligning the loop with 4 bytes and making the loop span in 2 32-byte chunks, however after your change, we no longer align the loop because of `
;; Skip alignment: 'Loop at G_M30565_IG06 PaddingNeeded= 11, MaxPadding= 8, LoopSize= 56, AlignmentBoundary= 16B.' `. With that, the loop spans 3 32-byte chunks and that could be the explaination of why we see regression.

![image](https://user-images.githubusercontent.com/12488060/121615900-c2ba1b80-ca16-11eb-8fa5-179c015e0972.png)


If I set `COMPlus_JitAlignLoopAdaptive=0`, we align the loop by adding `7 bytes` Before your change and non-adaptive alignment (on left below), we would align the loop by `4 bytes`. In both cases, the loop fits in 2 32-byte chunks.

![image](https://user-images.githubusercontent.com/12488060/121616406-b5516100-ca17-11eb-953e-60d5b07c3c48.png)

Finally, without loop alignment, `COMPlus_JitAlignLoops=0`, before and after changes, the loop takes 3 32-byte chunks to fit. 

![image](https://user-images.githubusercontent.com/12488060/121616763-622bde00-ca18-11eb-8666-2a43b6a52048.png)

To summarize, before your change, due to the loop alignment, we got better performance but after your change, we got the similar performance, had we didn't had loop alignment (modulo the performance impact of `align` instructions that is inside the 3rd nested loop).

**Comment by BruceForstall** [comment]
I don't understand your "To summarize" sentence.

I think we might be saying the same thing. In the "before my change" case, the inner loop ended up in 2 32-byte chunks. After, it ended up in 3 32-byte chunks. In the "after" case, setting `COMPlus_JitAlignLoopAdaptive=0` causes the inner loop to be in 2 32-byte chunks, giving the same performance as the "before" case. (Also, setting `COMPlus_JitAlignLoopAdaptive=0` in the "before" case gives worse performance, equivalent to my "after" case, due to the inner loop being in 3 32-byte chunks.)

**Comment by BruceForstall** [comment]
The `BenchmarksGame.FannkuchRedux_2` case is mysterious: the slow case includes 2 additional instructions in an un-executed path (loop cloning slow path), but all hot paths have the same instructions (some slightly different register allocation), and the hot inner loops are all properly aligned.

**Comment by BruceForstall** [comment]
The SciMark2 regression (both Windows and Linux) appears due to an additional spill / reload in the hot loop of `matmult`:

```
mov      r15d, dword ptr [rsp+24H] // here
cmp      r12d, r15d
mov      dword ptr [rsp+24H], r15d // here
```

This was fixed by https://github.com/dotnet/runtime/pull/53853.

**Comment by BruceForstall** [comment]
Since these have all been investigated, and follow-up issues opened, I'm closing this.
